### DISCUSSION TITLE
Better / More Powerful Dax Editor

### DISCUSSION THREAD

**Discussion by lulufredo** [TabularEditor/TabularEditor3#1157] [Ideas]
### Describe the solution you'd like

When batch-editing measures (using Script Dax) and such, it would be really useful to have more powerful editor features such as [multiple cursors](https://code.visualstudio.com/docs/editor/codebasics#_multiple-selections-multicursor) (including instance selection). Also the ability to [copy lines up / down](https://docs.posit.co/ide/user/ide/guide/productivity/text-editor.html#copy-lines-updown).

Note that these features exist in PBI's [DAX Formula bar](https://xxlbi.com/wp-content/uploads/2018/06/3MultipleSelections3.gif) (from [source](https://xxlbi.com/blog/dax-formula-bar-keyboard-shortcuts-in-power-bi-desktop/), search for "Multiple Selections" to see the actual shortcuts). But it's not as snappy as editing would be in TE3 (given the lag in PBI accessing different measures), and it's not possible to batch create/edit measures like it would be in the DAX Script Editor. 

### Goals

1. Ability to Select text instances quickly across a doc and place a cursor at each of them
2. Ability to place additional cursors vertically
3. Ability to duplicate lines vertically

### Background

I've tried find and replace but it's simply not as efficient.

**Comment by otykier** [comment]
1. This is already possible, but there is no key binding for this action by default. Go to **Tools > Preferences > Keyboard**. Look for the “Editor.MultipleSelectAddEach” and “Editor.MultipleSelectAddNext” commands. [More info here](https://data-goblins.com/power-bi/tabular-editor-keyboard-shortcuts#:~:text=SHORTCUTS%20FOR%20TEXT,could%20hypothetically%20configure%3A).
2. Could you clarify exactly what you mean by this?
3. Duplicating the current line is also possible. The default shortcut is CTRL+D. You can use ALT+Up/Down arrow to shift the current line up or down.

**Reply by lulufredo** [reply]
Thanks so much for the quick reply—I'll take a look!

For #2, it's [this](https://code.visualstudio.com/assets/docs/editor/codebasics/multicursor.gif)—no instance based but just cursors added straight up or down

**Reply by otykier** [reply]
I see. I’m afraid we don’t have a shortcut for selecting an entire identical block vertically, but you can make rectangular selections by holding down ALT and SHIFT and then using the arrow keys or the mouse.

**Reply by lulufredo** [reply]
Amazing—thanks so much!

**Reply by lulufredo** [reply]
Actually there is one missing (just realized), not sure if I should submit a separate ticket: Unselect instance. It's [Ctrl+U](https://stackoverflow.com/questions/58858437/vscode-deselect-one-from-change-all-occurrences) in VSCode.
